System/Win32/MinTTY.hsc

```diff
@@ -0,0 +1,207 @@
+--
+-- Maintainer  :  Esa Ilari Vuokko <ei@vuokko.info>
+-- Stability   :  provisional
+-- Portability :  portable
+--
+-- A function to check if the current terminal uses MinTTY.
+-- Much of this code was originally authored by Phil Ruffwind.
+--
+-----------------------------------------------------------------------------
+
+module System.Win32.MinTTY (isMinTTY) where
+
+import Graphics.Win32.Misc
+import System.Win32.DLL
+import System.Win32.File
+import System.Win32.Types
+
+#if MIN_VERSION_base(4,6,0)
+import Control.Exception (catch)
+#endif
+import Control.Monad (void)
+import Data.List (isInfixOf, isSuffixOf)
+import Foreign hiding (void)
+import Foreign.C.String
+import Foreign.C.Types
+
+#if __GLASGOW_HASKELL__ < 711
+#let alignment t = "%lu", (unsigned long)offsetof(struct {char x__; t (y__); }, y__)
+#endif
+
+##include "windows_cconv.h"
+#include <windows.h>
+
+-- | Returns 'True' is running on a MinTTY console (e.g., Cygwin or MSYS).
+-- Returns 'False' otherwise.
+isMinTTY :: IO Bool
+isMinTTY = do
+    h <- getStdHandle sTD_ERROR_HANDLE
+    fileType <- getFileType h
+    if fileType /= fILE_TYPE_PIPE
+      then pure False
+      else do
+        -- GetFileNameByHandleEx is only available on Vista and later, so if
+        -- we're on an older version of Windows, we must default to using
+        -- NtQueryObject.
+        vistaPlus <- isVistaOrLater
+        if vistaPlus
+           then isMinTTYVista h
+           else isMinTTYCompat h
+
+isMinTTYVista :: HANDLE -> IO Bool
+isMinTTYVista h = do
+    fn <- getFileNameByHandle h
+    pure $ cygwinMSYSCheck fn
+  `catch` \ (_ :: IOError) ->
+    pure False
+
+isMinTTYCompat :: HANDLE -> IO Bool
+isMinTTYCompat h = do
+    fn <- ntQueryObjectNameInformation h
+    pure $ cygwinMSYSCheck fn
+  `catch` \ (_ :: IOError) ->
+    pure False
+
+cygwinMSYSCheck :: String -> Bool
+cygwinMSYSCheck fn = ("\\cygwin-" `isInfixOf` fn || "\\msys-" `isInfixOf` fn) &&
+            "-pty" `isInfixOf` fn &&
+            "-master" `isSuffixOf` fn
+-- Note that GetFileInformationByHandleEx might return something like:
+--
+--    \msys-dd50a72ab4668b33-pty1-to-master
+--
+-- But NtQueryObject might return something like:
+--
+--    \Device\NamedPipe\msys-dd50a72ab4668b33-pty1-to-master
+--
+-- This means we can't rely on "cygwin-" or "msys-" being at the very start,
+-- so we check for their presence using `isInfixOf` instead of `isPrefixOf`.
+
+getFileNameByHandle :: HANDLE -> IO String
+getFileNameByHandle h = do
+  let sizeOfDWORD = sizeOf (undefined :: DWORD)
+  let sizeOfWchar = sizeOf (undefined :: CWchar)
+  -- note: implicitly assuming that DWORD has stronger alignment than wchar_t
+  let bufSize = sizeOfDWORD + mAX_PATH * sizeOfWchar
+  allocaBytes bufSize $ \buf -> do
+    getFileInformationByHandleEx h fILE_NAME_INFO buf (fromIntegral bufSize)
+    len :: DWORD <- peek buf
+    let len' = fromIntegral len `div` sizeOfWchar
+    peekCWStringLen (buf `plusPtr` sizeOfDWORD, min len' mAX_PATH)
```

Comment:
This (and `ntQueryObjectNameInformation` below) do a lot of memory manipulation which, as far as I can see, making some assumptions about how the memory being written to `buf` is laid out. We have to do this since the structs that we need (`FILE_NAME_INFO` and `OBJECT_NAME_INFORMATION`, respectively) don't appear to be available on the version of MinGW-w64 that ship with GHC (8.0.1, at least).

One alternative would be conditionally defining these structs in a header file (if they aren't already defined), and rely on the FFI to marshal them via `Storable` instances.